packages/captp/src/atomics.js

```diff
@@ -0,0 +1,149 @@
+  );
+  const lenbuf = new BigUint64Array(transferBuffer, 0, 1);
+  const statusbuf = new Int32Array(transferBuffer, lenbuf.byteLength, 1);
+  const overheadLength = lenbuf.byteLength + statusbuf.byteLength;
+  assert.equal(
+    overheadLength,
+    TRANSFER_OVERHEAD_LENGTH,
+    X`Internal error; actual overhead ${overheadLength} of bytes is not TRANSFER_OVERHEAD_LENGTH ${TRANSFER_OVERHEAD_LENGTH}`,
+  );
+  const databuf = new Uint8Array(
+    transferBuffer,
+    lenbuf.byteLength + statusbuf.byteLength,
+  );
+  assert(
+    databuf.byteLength >= MIN_DATA_BUFFER_LENGTH,
+    X`Transfer buffer of size ${transferBuffer.byteLength} only supports ${databuf.byteLength} data bytes; need at least ${MIN_DATA_BUFFER_LENGTH}`,
+  );
+  return { statusbuf, lenbuf, databuf };
+};
+
+/**
+ * Create a trapHost that can be paired with makeAtomicsTrapGuest.
+ *
+ * This host encodes the transfer buffer and returns it in consecutive slices
+ * when the guest iterates over it.
+ *
+ * @param {SharedArrayBuffer} transferBuffer
+ * @returns {TrapHost}
+ */
+export const makeAtomicsTrapHost = transferBuffer => {
+  const { statusbuf, lenbuf, databuf } = splitTransferBuffer(transferBuffer);
+
+  const te = new TextEncoder();
+
+  return async function* trapHost([isReject, serialized]) {
+    // Get the complete encoded message buffer.
+    const json = JSON.stringify(serialized);
+    const encoded = te.encode(json);
+
+    // Send chunks in the data transfer buffer.
+    let i = 0;
+    let done = false;
+    while (!done) {
+      // Copy the next slice of the encoded arry to the data buffer.
+      const subenc = encoded.subarray(i, i + databuf.length);
+      databuf.set(subenc);
+
+      // Save the length of the remaining data.
+      lenbuf[0] = BigInt(encoded.length - i);
+
+      // Calculate the next slice, and whether this is the last one.
+      i += subenc.length;
+      done = i >= encoded.length;
+
+      // Find bitflags to represent the rejected and finished state.
+      const rejectFlag = isReject ? STATUS_FLAG_REJECT : 0;
+      const doneFlag = done ? STATUS_FLAG_DONE : 0;
+
+      // Notify our guest for this data buffer.
+
+      // eslint-disable-next-line no-bitwise
+      statusbuf[0] = rejectFlag | doneFlag;
+      Atomics.notify(statusbuf, 0, +Infinity);
+
+      // Wait until the next call.
+      yield;
+    }
+  };
+};
+
+/**
+ * Create a trapGuest that can be paired with makeAtomicsTrapHost.
+ *
+ * This guest iterates through the consecutive slices of the JSON-encoded data,
+ * then returns it.
+ *
+ * @param {SharedArrayBuffer} transferBuffer
+ * @returns {TrapGuest}
+ */
+export const makeAtomicsTrapGuest = transferBuffer => {
+  const { statusbuf, lenbuf, databuf } = splitTransferBuffer(transferBuffer);
+
+  return ({ startTrap }) => {
+    const td = new TextDecoder('utf-8');
+
+    let json = '';
```

Comment:
> There's an odd asymmetry between the host and guest.
> * The host does a TextEncoder.encode all at once, into a utf8 binary.
> Guest
> * For each one, it does a TextDecoder.decode, to turn it into successive strings that it appends to the `json` string variable.

I'm taking advantage of the `{ stream: true }` option to the TextDecoder.decode.  That preserves any partially-decoded state between successive calls to decode (https://developer.mozilla.org/en-US/docs/Web/API/TextDecoder/decode#syntax).  So, there's no problem.

I will add a comment indicating this.

Reply:
I see. Good to know. Two things:

XS implements (plans to implement?) a subset of TextDecoder. Does this subset support this streaming feature? (attn @phoddie )

Is there any advantage to doing it this way, rather than in the say that's more obviously symmetric and correct?

Reply:
The advantage is not needing to allocate (and hence either track or throw away) intermediate decoding buffers.

Reply:
But you're right, the Moddable TextDecoder doesn't seem to do stream mode.  I'll rewrite accordingly, throwing away the intermediate buffers.

Reply:
This is now fine. Resolving.

But I do want to suggest one possible improvement: Since we know before the loop how big the buffer needs to be, can we just preallocate the whole thing, and then incrementally write into it until we're done?

Reply:
> But I do want to suggest one possible improvement: Since we know before the loop how big the buffer needs to be, can we just preallocate the whole thing, and then incrementally write into it until we're done?

We only learn the length after the first `Atomics.wait` (which is in the loop), so we can't preallocate it.